src/stream/src/executor/upstream_sink_union.rs

```diff
@@ -311,25 +422,79 @@ impl UpstreamSinkUnionExecutor {
                 "UpstreamSinkUnion",
             ]);
 
-        let upstreams = DynamicReceivers::new(inputs, Some(barrier_align), None);
+        let upstreams = DynamicReceivers::new(inputs, Some(barrier_align.clone()), None);
         pin_mut!(upstreams);
 
-        while let Some(msg) = upstreams.next().await {
-            yield msg?;
+        let mut current_barrier = None;
+
+        // Here, `tokio::select` cannot be used directly in `try_stream` function.
+        // Err for breaking the loop. Ok(None) for continuing the loop.
+        let mut select_once = async || -> StreamExecutorResult<Option<Message>> {
+            tokio::select! {
+                biased;
+
+                Some(msg) = upstreams.next() => {
+                    let msg = msg.context("UpstreamSinkUnionExecutor pull upstream failed")?;
+                    if let Message::Barrier(barrier) = &msg {
+                        let current_barrier = current_barrier.take().unwrap();
+                        assert_equal_dispatcher_barrier(&current_barrier, barrier);
+                        self.handle_update(&mut upstreams, actor_id, &barrier_align, barrier).await?;
+                    }
+                    Ok(Some(msg))
+                }
+
+                Some(barrier) = self.barrier_rx.recv(), if current_barrier.is_none() => {
+                    // Here, if there's no upstream, we should process the barrier directly and send it out.
+                    // Otherwise, we need to forward the barrier to the upstream and then wait in the first branch
+                    // until the upstreams have processed the barrier.
+                    if upstreams.is_empty() {
+                        self.handle_update(&mut upstreams, actor_id, &barrier_align, &barrier).await?;
+                        Ok(Some(Message::Barrier(barrier.clone())))
+                    } else {
+                        for tx in self.barrier_tx_map.values() {
+                            tx.send(barrier.clone())
+                                .map_err(|e| StreamExecutorError::from(anyhow::anyhow!(e)))?;
+                        }
+                        current_barrier = Some(barrier.clone());
+                        Ok(None)
+                    }
+                }
+
+                else => {
```

Comment:
In current implementation, will `upstreams.next()` return `None` when there is no upstream? If so, will it be incorrectly treated as error in this case?